### ISSUE TITLE
Support OverloadedRecordDot

### ISSUE THREAD

**Issue by deemp** [well-typed/large-records#128]
Hi! I'd like to use your`large-anon`  library with `GHC 9.2.4` but it seems to be inconvenient. 

My example that works with `GHC 9.0.2`:

```hs
{- FOURMOLU_DISABLE -}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE OverloadedLabels #-}
{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
{-# OPTIONS_GHC -fplugin=RecordDotPreprocessor #-}
{-# OPTIONS_GHC -fplugin=Data.Record.Anon.Plugin #-}
{- FOURMOLU_ENABLE -}

module Main (main) where

import Data.Record.Anon
import Data.Record.Anon.Simple (Record, empty, insert)
import qualified Data.Record.Anon.Simple as Anon

main :: IO ()
main = print 3

magenta :: Record ["red" := Double, "green" := Double, "blue" := Double]
magenta = ANON{red = 1, green = 0, blue = 1}

purple :: Record ["red" := Double, "green" := Double, "blue" := Double]
purple =
  insert #red 0.5 $
    insert #green 0 $
      insert #blue 0.5 $
        empty

reduceRed :: RowHasField "red" r Double => Record r -> Record r
reduceRed c = c{red = c.red * 0.9}
```

The problem is that this `GHC` doesn't support `OverloadedRecordDot` extension, so `fourmolu` makes bad refactorings.

So, I try to switch to `GHC 9.2.4` and get errors when using `HLS` and `fourmolu`. `HLS` doesn't recognize `ANON` from the example, and `fourmolu` doesn't respect the plugin `RecordDotPreprocessor`.

Did I do something wrong? Are you planning to support `large-anon` with `OverloadedRecordDot` extension instead of `RecordDotPreprocessor`?

**Comment by CrystalSplitter** [comment]
Trying to debug my own problem with large-anon, I came across your issue. I have however gotten `ANON` to work, at least in `GHC 9.2.5`. Perhaps my minimum `ANON` set up may help?

```hs
{-# LANGUAGE DataKinds        #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE TypeApplications #-}
{-# OPTIONS_GHC -fplugin=RecordDotPreprocessor #-}
{-# OPTIONS_GHC -fplugin=Data.Record.Anon.Plugin #-}

module MyLib where

import           Data.List (foldl')

func :: Int -> Int
func end = res.y
  where
    res = foldl' go (ANON {x = 1, y = 1}) [1..end]
    go st i = ANON
        { x = x'
        , y = st.y + x'
        }
      where
        x' = st.x + st.y + i

```

My cabal build-depends were:

```hs
build-depends:    base ^>=4.16.4.0
                  , large-anon ^>=0.1.1
                  , record-hasfield ^>=1.0
                  , record-dot-preprocessor ^>= 0.2.15
```

**Comment by deemp** [comment]
Thanks, @CrystalSplitter! Also got your setup working. In fact, on can have at the top of the file

```hs
-- {-# LANGUAGE OverloadedRecordDot #-}
```

And uncomment it before formatting, then comment again. 

These actions can be automated and distributed as a VSCode extension ([example](https://github.com/deemp/vscode-purs-keybindings)).

**Comment by edsko** [comment]
@deemp , @CrystalSplitter , I believe this is now solved in #134 . If you want to use `large-anon` with record dot syntax and Fourmolu, you now have two choices, both of which will require the use of ghc 9.2 or up:

- Enable `OverloadedRecordDot`, and still use the `RecordDotPreprocessor` plugin. If you do this, you will be limited to simple record updates; something like `r{a.b = True}` will be rejected by `fourmolu` (but `r{a = True}` will be fine, since this syntax is regular syntax, even if `RecordDotPreprocessor` then interprets it in a different way).
- Enable `OverloadedRecordDot` as well as `OverloadedRecordUpdate`, and then don't rely on `RecordDotPreprocessor` at all anymore. If you want to go this route, you will also need to enable `RebindableSyntax`, which is currently a requirement for `OverloadedRecordUpdate`. `RebindableSyntax` has some repercussions about what it in scope, but `large-anon` now offers a new module `Data.Record.Anon.Overloading`; if you import this, you should not have any unwanted consequences from using `RebindableSyntax`.

So, option (1) looks like

```haskell
{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE TypeApplications    #-}
{-# LANGUAGE TypeOperators       #-}

{-# OPTIONS_GHC -fplugin=RecordDotPreprocessor -fplugin=Data.Record.Anon.Plugin #-}

module Foo where

import Data.Record.Anon
import Data.Record.Anon.Simple
```

at which point you can use `large-anon` as usual, except that you have no nested record updates; option (2) looks like

```haskell
{-# LANGUAGE DataKinds              #-}
{-# LANGUAGE OverloadedRecordDot    #-}
{-# LANGUAGE OverloadedRecordUpdate #-}
{-# LANGUAGE RebindableSyntax       #-}
{-# LANGUAGE TypeApplications       #-}
{-# LANGUAGE TypeOperators          #-}

{-# OPTIONS_GHC -fplugin=Data.Record.Anon.Plugin #-}

module Foo where

import Data.Record.Anon
import Data.Record.Anon.Overloading
import Data.Record.Anon.Simple
```

There is no difference between ghc 9.2 and 9.4; the `OverloadedRecordUpdate` extension is still work in progress, despite what is claimed in that Fourmolu ticket you cited. If you _do_ want to use ghc 9.4, and you want to use the preprocessor, you will need a new release, a PR that I wrote against `record-dot-processor` to support 9.4 just got merged but isn't yet released, though I expect that will happen soon.
